### PR TITLE
Support verifying a sender using messagingToken

### PR DESCRIPTION
**Pull request by Enriqe**
Currently, `WindowPortEmulator ` inside `messaging.js` [compares](https://github.com/ampproject/amphtml/blob/87d71fc4ee87622ae4c0591bff6ba39f82f1f615/extensions/amp-viewer-integration/0.1/messaging/messaging.js#L82) the `event.origin` to its `this.origin_` set by the `amp-viewer-integration` code:

https://github.com/ampproject/amphtml/blob/471bbb957ba9ee4c3d1778dbb987d8fee3352e0b/extensions/amp-viewer-integration/0.1/amp-viewer-integration.js#L87

https://github.com/ampproject/amphtml/blob/471bbb957ba9ee4c3d1778dbb987d8fee3352e0b/extensions/amp-viewer-integration/0.1/amp-viewer-integration.js#L122-L124

This causes issues when the parent window of the viewer does not have an origin (#28225).

This PR will allow the `amp-story-player` to send a `messagingToken` in the `waitForHandshakeFromDocument ` method when establishing a handshake with the STAMP. Paired with the `#messagingToken=storyPlayerToken` parameter set on the STAMP's url, the `WindowPortEmulator` can then use those to verify the communication instead of comparing the `origin`s of the window and event.

https://github.com/ampproject/amphtml/blob/471bbb957ba9ee4c3d1778dbb987d8fee3352e0b/src/amp-story-player/amp-story-player-impl.js#L241-L245

### CODE REVIEW

**Comment by Enriqe** [comment]
@zhangsu since I know you implemented #20634 to introduce a secure way to embed a sandboxed iframe, I was wondering if you could review this PR? Hoping I'm not introducing a vulnerability of the attack described in https://github.com/ampproject/amphtml/pull/20634#pullrequestreview-199247853 by @choumx with this PR

**Comment by choumx** [comment]
Security review should come from our ISE consult @molnarg.

File: extensions/amp-viewer-integration/0.1/messaging/messaging.js [hunk 1]
```diff
@@ -79,7 +82,11 @@ export class WindowPortEmulator {
    */
   addEventListener(eventType, handler) {
     this.win_.addEventListener('message', (event) => {
-      if (event.origin == this.origin_ && event.source == this.target_) {
+      if (
+        (event.data.messagingToken === this.token_ ||
+          event.origin == this.origin_) &&
```
Comment (fstanis) [inline extensions/amp-viewer-integration/0.1/messaging/messaging.js:87]: Just a small suggestion: if I understand correctly, this means having a messaging token completely disables the origin checking behaviour.

Would it be possible to keep both (i.e. change `||` to `&&` here)? Then you can just use `'null'` as the origin parameter when you create a `WindowPortEmulator` when there's no origin (this is what we do for the AMP for Email viewer).

Comment (Enriqe) [reply extensions/amp-viewer-integration/0.1/messaging/messaging.js:87]: I actually tried passing `'null'` as the origin parameter https://github.com/ampproject/amphtml/issues/28225#issuecomment-625461308, but the problem is that when the `WindowPortEmulator` receives a message from the viewer's parent window, `event.origin` actually has the `origin` of the parent window, making this check fail.

Comment (molnarg) [reply extensions/amp-viewer-integration/0.1/messaging/messaging.js:87]: token_ is null if it was not specified, so a message with messagingToken=null could bypass the origin check, if the token was not expected.

Btw, other token checking is done in the Messaging class, not sure how this interacts with that, would be better to have all this logic in a single class. Also, the Messaging class implemented this check in a better way here without the flaw mentioned above: https://github.com/ampproject/amphtml/blob/b65705aae8892fe3c4dceab624296fbfbee58315/extensions/amp-viewer-integration/0.1/messaging/messaging.js#L301

**Comment by molnarg** [comment]
See https://github.com/ampproject/amphtml/issues/28225#issuecomment-629155166, maybe we won't need this.

**Pull request closed without merging**